### ISSUE TITLE
Run inline tests in parallel

### ISSUE THREAD

**Issue by bcc32** [ocaml/dune#1516]
ppx_inline_test supports labeling inline tests in a given module with a "partition" and only running tests from a specified partition, meaning tests can be run in parallel.  Dune supports running inline tests from different libraries in parallel, but it would be nice to get parallelism at the module level as well.

**Comment by diml** [comment]
Indeed. We planned to do this initially but never got to implement it. This will get implemented at some point, though if you want to give it a try now, I'm happy to point you out to the code to update.

**Comment by hhugo** [comment]
Running partitions in parallel would also allow dune to report all errors in all files at once.
I think ppx_inline_test currently fail after the first partition if there is a failure. This should probably be fixed in ppx_inline_test though. @jeremiedimino any opinion ?

**Comment by jeremiedimino** [comment]
We could change `ppx_inline_test` to report all errors, however this raise another issue. You don't see any error until all the tests have finished. This can be annoying if there are some long running tests. We changed the way we run inline tests at Jane Street last year and users raised this issue.

So changing Dune to run inline tests in parallel seems more promising to me.

**Comment by hhugo** [comment]
To give a bit more details.
inline tests using `ppx_inline_tests`/`ppx_expect` can list partitions with the `-list-partition` flag. Each partition can run individually with the `-partition <name>` flag.
To me it seems that the most annoying part is make that behavior configuration in inline tests backend stanza. 

See how it was implemented in jenga https://github.com/janestreet-deprecated/jenga-rules/blob/1ded5e85138e342126a8807865475846eba9d405/lib/root.ml#L4704

**Comment by hhugo** [comment]
>> Running partitions in parallel would also allow dune to report all errors in all files at once.
I think ppx_inline_test currently fail after the first partition if there is a failure. This should probably be fixed in ppx_inline_test though. @jeremiedimino any opinion ?

> We could change `ppx_inline_test` to report all errors

Looking at this again, the issue I was having is ppx_expect reporting errors one file at a time.
It turns out that with the current setup, ppx_expect just writes corrected file on disk and let dune do the diffing.
The diffing is done is sequence by dune, stopping at the first diff.
```ocaml
           let run_tests = Action.chdir (Path.build dir) action in
           Action.Full.make ~sandbox
           @@ Action.progn
                (run_tests
                :: List.map source_files ~f:(fun fn ->
                       Action.diff ~optional:true fn
                         (Path.Build.extend_basename
                            (Path.as_in_build_dir_exn fn)
                            ~suffix:".corrected")))))
```

**Comment by hhugo** [comment]
I've tried to implement the feature and got stuck realizing that I want to declare an arbitrary number of alias action (`add_alias_action`) based on a list of partition previously computed. I think this cannot currently be done, is it correct ?

**Comment by Alizter** [comment]
I thought that #6933 would address this issue, however I was mistaken. What it does address is a separate issue I have submitted as https://github.com/ocaml/dune/issues/6942.

@hhugo Rather than messing around with alias', could you rebase your fix on top of #6933 and see if the `(conc )` (`progn`-like concurrency action) allows you to run your actions together? You should probably run the test and diff in the same action and `(conc)` all of those.

**Comment by rgrinberg** [comment]
> I've tried to implement the feature and got stuck realizing that I want to declare an arbitrary number of alias action (`add_alias_action`) based on a list of partition previously computed. I think this cannot currently be done, is it correct ?

What problems did you run into? This seems like the right approach.

**Comment by rgrinberg** [comment]
@hhugo did you have a syntax in mind for allowing the user to partition their library through the dune file?

**Comment by hhugo** [comment]
I've been looking at adding an optional field to the inline_tests backend  (e.g. `list_partitions_flags`).
When present, the executable would be executed with the given flags to retrieve the list of partitions.

Is here my WIP https://github.com/hhugo/dune/commit/30f847dbf8e2e08d32bfedd94a983d063f4f2770

**Comment by hhugo** [comment]
> I thought that #6933 would address this issue, however I was mistaken. What it does address is a separate issue I have submitted as #6942.
> 
> @hhugo Rather than messing around with aliases, could you rebase your fix on top of #6933 and see if the `(conc )` (`progn`-like concurrency action) allows you to run your actions together? You should probably run the test and diff in the same action and `(conc)` all of those.
> 
> If it works, then that is more evidence for #6933 which we can include in the discussion in #6937.

I've opened #7012
